sympy/functions/elementary/exponential.py

```diff
@@ -693,22 +693,30 @@ def as_real_imag(self, deep=True, **hints):
         else:
             return (log(abs), arg)
 
+    @property
+    def _arg_is_1(self):
+        from sympy.physics.units import Quantity, Dimension
+        if self.has(Quantity, Dimension):
```

Comment:
This sort of thing is disconcerting. Will everything in SymPy have to have guards against Quantity? 

I guess if we decide that `log` is really the only function that supports quantities, then it isn't a big deal.

Reply:
> This sort of thing is disconcerting

I know. I think this is a special situation for `log` because it is 0 at arg=1. So the assumptions test to see if it is 1 or not by computing `arg - 1`. An alternative would be to use a try/except to handle that error. Maybe that would be preferred to this current proposal?

Reply:
like this:
```
try:
 arg - 1
except ValueError as e:
 if 'incompatible' not in e.args[0]:
  raise e
```

Or should the error match be exact, `if 'summation of quantities of incompatible dimensions' != e.args[0]`?

Reply:
It's still a problem that the core has to know about Quantity. Even if we abstract away Quantity, it complicates the code everywhere if we can't even assume that expressions can be added to scalars.

Reply:
I think we may run into problems with other functions, too, such as trigonometric functions etc.
Maybe a totally crazy idea, but to avoid the burden of checking for quantities and units all the time, why don't we switch this off as default and just add a function `check_dimensions` to allow the user to check the dimensions of any expression at any point in time? This way, users could do whatever algebra they like with quantities, but in the end they will not get a meaningful result if the dimensions are not consistent. It would then be up to the user to algebraically resolve any units in functions to something meaningful before they perform a calculation: `log(1*meter)` would be unevaluated unless the user e.g. takes the difference: `log(1*meter) - log(1*centimeter) = log(1/0.01) = 4.6`

Reply:
In the interim, could we not add error handling to all of the `_eval_is_...` functions? They should only return `True` if the condition is actually true, whereas if it is not true or raises an error, as in the case of Quantities, the result should simply be `False`.